scripts/release/download-cabal-install-release-binaries.sh

```diff
@@ -0,0 +1,42 @@
+#!/bin/bash
+
+# A script to download the release binary files for a given cabal release
+# from upstream "downlods.haskell.org".
+# It accepts the first and only argument as the release number.
+#
+# useage:-
+#   $ download-cabal-install-release-binaries.sh "3.10.1.0"
+#
+# This was initally made to be used with ./create-release-metadata-for-ghcup.sh
+
+set -eu
+set -o pipefail
+
+RELEASE=$1
+
+echo "RELEASE: $RELEASE"
+
+for com in wget sha256sum ; do
+        command -V ${com} >/dev/null 2>&1
+done
+
+[ ! -d "binary-downloads/cabal-install-${RELEASE}-binaries" ]
+
+mkdir -p "binary-downloads/cabal-install-${RELEASE}-binaries"
+
+cd "binary-downloads/cabal-install-${RELEASE}-binaries"
+
+## Download release files
+echo "Downloading form: \"https://downloads.haskell.org/~cabal/cabal-install-${RELEASE}/\""
+wget --no-parent -r --reject "index.html*" --no-directories "https://downloads.haskell.org/~cabal/cabal-install-${RELEASE}/"
```

Comment:
This is the wrong way around and would mean we download twice: once manually and once after we uploaded (why?).

The artifacts are built on gitlab CI. The release manager will have to download them to their local machine in order to upload them to downloads.haskell.org. And they want to perform the yaml step after they downloaded from gitlab, not after they uploaded to the download server.

Reply:
> I see. Since we don't seem to do github binary releases for cabal-install, i suppose i'd have to download them from upstream links like https://downloads.haskell.org/~cabal/cabal-install-3.10.1.0/ ?

could've mentioned here when i asked to confirm.

What's the correct url here?

Reply:
@chreekat might know

Reply:
🕯️ * summons @chreekat * 🕯️

Reply:
👻

Reply:
You can drop the ~ and just use https://downloads.haskell.org/cabal/ as the base url. Otherwise that's right. And if you don't want maintainers to have to download things locally, once the script is in good shape we can add it to the release pipeline at gitlab.haskell.org/Haskell/cabal. I can help with that.

Reply:
> You can drop the ~ and just use https://downloads.haskell.org/cabal/ as the base url. 

That doesn't work. We need to download from the gitlab CI pipeline that creates the release, because we have to anyway in order to upload it to downloads.haskell.org.

Reply:
Oh duh. Well I think a release manager would be better qualified to answer. :) 

I'm happy to help automate later at any rate!

Reply:
> Oh duh. Well I think a release manager would be better qualified to answer. :)

We're asking for the URL to download artifacts from gitlab.

----

Anyway, @arjunkathuria look here: https://docs.gitlab.com/ee/api/job_artifacts.html#download-the-artifacts-archive

An example is:

```sh
curl -L 'https://gitlab.haskell.org/api/v4/projects/1164/jobs/artifacts/cabal-install-v3.10.2.1/download?job=build-aarch64-darwin' -o foo.zip
```

I'm not sure what the correct job names for the matrix builds are... this downloads a zip archive with all the individual artifacts in it that a job produces. There's also a way to just download a specific artifact from a job.

One major caveat: this only works if the pipeline as a whole succeeded.

Reply:
This gets a bit more involved.

To get the artifacts for all platforms and architectures, from what i can read from the docs, we'd either have to give:

 
- individual job ids for each one to the script.
or
- hardcode all the job names like `build-x86_64-windows`, `build-linux: [x86_64, x86_64-linux, deb10] ` and a branch name like `main`.

unzip the `artifact.zip` archives and put the `cabal-install-*` files in a single directory.

and even then it'll work only if the pipeline succeeds for all architectures and platforms at once.

How does this happen currently, manual by-hand downloads by the release manager?

Reply:
Yes, it's a little more work.

Reply:
Do you think that this is better suited as its own issue/PR pair? 
The scope of this particular PR was to make a script that generates metadata for GHCup for a cabal release.

I think we can iterate on the how-to and where-to download parts that make releases easier as its own rel-eng issue as it would involve into its own discussion/interface requirements.

Reply:
> We're asking for the URL to download artifacts from gitlab.

That's why I suggest talking to a past/current Cabal release manager. I imagine they already have a workflow for this. While I can point to release pipelines and explain in general how to find the artifacts, it would be better to know what the actual current process is. Maybe @Kleidukos can help.

> Do you think that this is better suited as its own issue/PR pair?
The scope of this particular PR was to make a script that generates metadata for GHCup for a cabal release.

Yeah, I totally agree with that.

To actually automate it, it would be a lot easier to run the script in the GitLab pipeline that generates the bindists in the first place. That's what I can help with. https://github.com/haskell/cabal/blob/master/.gitlab-ci.yml is the entrypoint.

Reply:
I think the best course of action is to purge gitlab CI.

Reply:
You may want to re-purpose the tools of the GHC release engineering team: 

* https://gitlab.haskell.org/ghc/ghc/-/tree/master/.gitlab/rel_eng/mk-ghcup-metadata?ref_type=heads

* https://gitlab.haskell.org/ghc/ghc/-/tree/master/.gitlab/rel_eng/fetch-gitlab-artifacts?ref_type=heads

Reply:
Yea, they do it in two steps.

1. get the job ids from a pipeline id.
2. try to get artifacts from these job ids and error if anyone of those failed.
3. If successful, generate metadata from these artifacts.

i suppose i try repurposing the logic in the shell script here. 
If that's not required or if that needs to be a separate issue/PR pair, please do let me know.